\documentclass{article}
\input{../config/commontex}

\title{Release notes}
\author{Lobry, J.R. \and Nec\c{s}ulea, A. \and Palmeira, L. \and Penel, S.}

\usepackage{Sweave}
\begin{document}
%
% To change the R input/output style:
%
\definecolor{Soutput}{rgb}{0,0,0.56}
\definecolor{Sinput}{rgb}{0.56,0,0}
\DefineVerbatimEnvironment{Sinput}{Verbatim}
{formatcom={\color{Sinput}},fontsize=\footnotesize, baselinestretch=0.75}
\DefineVerbatimEnvironment{Soutput}{Verbatim}
{formatcom={\color{Soutput}},fontsize=\footnotesize, baselinestretch=0.75}
%
% This removes the extra spacing after code and output chunks in Sweave,
% but keeps the spacing around the whole block.
%
\fvset{listparameters={\setlength{\topsep}{0pt}}}
\renewenvironment{Schunk}{\vspace{\topsep}}{\vspace{\topsep}}
%
% Rlogo
%
\newcommand{\Rlogo}{\protect\includegraphics[height=1.8ex,keepaspectratio]{../figs/Rlogo.pdf}}
%
% Shortcut for seqinR:
%
\newcommand{\seqinr}{\texttt{seqin\bf{R}}}
\newcommand{\Seqinr}{\texttt{Seqin\bf{R}}}
\fvset{fontsize= \scriptsize}
%
% R output options and libraries to be loaded.
%
%
%  Sweave Options
%
% Put all figures in the fig folder and start the name with current file name.
% Do not produce EPS files
%


\maketitle
% BEGIN - DO NOT REMOVE THIS LINE

\section*{Introduction}

The release notes are listed in reverse chronological order:
most recent on top.

\section*{2.0 series}

\subsection*{release 2.0-7}

\begin{itemize}

\item A new utility function \texttt{where.is.this.acc()} was introduced
to loop over all availabale ACNUC databases to look for
a given sequence accession number. This is useful when you have
a sequence accession number and you don't know in which database
it is present. The documentation of the function \texttt{choosebank()}
was also changed to make a link to this function. As suggested by 
Avril Coghlan, the function has an argument \texttt{stopAtFirst} defaulting
to \texttt{TRUE} that stops the search at the first database found
with the given accession number.

\item As pointed out 05 Nov 2009 by Darren Obbard on the seqinr diffusion list
the argument \texttt{forceToLower = FALSE} in function \texttt{comp()}
was not honored. This is now fixed and a new sanity check was added in
the example section of the documentation of the function.

\item Documentation for the function \texttt{uco()} for codon usage
table computation was updated with new bibliographical 
references \cite{lobrychessel, SuzukiH2008}.

\item As basic regular expressions are defunct since R 2.11, the
\texttt{extended} argument in functions \texttt{words.pos()} and
\texttt{trimSpace()} was no more necessary. It is now deleted. 

\end{itemize}

\subsection*{release 2.0-6}

\begin{itemize}

\item The old argument \texttt{File} in function \texttt{read.fasta()}
that was deprecated since release 1.1-3 is no more valid.
Just use \texttt{file} instead.

\item New function \texttt{stutterabif()} to estimate stutter ratio.

\item Function \texttt{plotabif()} has a new default value for its
\texttt{ylim} argument: \texttt{c(min(y), max(y))} now instead
of \texttt{c(0, max(y))} previously to help ploting data with
a highly negative baseline.

\item Function \texttt{peakabif()} now returns in addition an estimate
of the baseline value.

\item New utility \texttt{baselineabif()} to estimate the baseline
value.

\item There was time shift of one datapoint unit for the peak locations
returned by the \texttt{peakabif()} function, this is now fixed
and the documentation is more explicit for the units used.

\item New utility function \texttt{fastacc()} to compute the number
of alleles in common between a genetic profile and a database of
genetic profiles.

\end{itemize}

\subsection*{release 2.0-5}

\begin{itemize}

\item New utility function \texttt{circle()} to draw a circle.

\item Two more examples of files to be imported with the \texttt{readBins()}
and \texttt{readPanels()} functions are now available in the \texttt{abif}
folder: \texttt{NGM\_Bins.txt} and \texttt{NGM\_Pa.txt}, respectively.

\item New function \texttt{plotPanels()} to plot amplicon size ranges of STR
kits data.

\item New utility function \texttt{col2alpha()} to add a transparency 
chanel to a standard R color.

\item New ABIF example file \texttt{samplefsa2ps.fsa} used in the
\texttt{read.abif()} function to reproduce figure 1A from \cite{Krawczyk2009}.

\item New function \texttt{move()} aliased as \texttt{mv()} to rename
an object without deep copy.

\item New function \texttt{swap()} to exchange two objects.

\end{itemize}

\subsection*{release 2.0-4}

\begin{itemize}

\item Configuration files to be imported by the \texttt{readBins()}
function may have trailling tabulations, as for instance in the
test file \texttt{Prototype\_PowerPlex\_EP01\_Bins.txt} for allele
9 at locus \texttt{D3S1358} and for allele 14 at locus \texttt{D12S391}.
This was a source of trouble during importation. This is now fixed
and the above mentioned file is used as a quality control. A warning
is now issued if the number of columns in the data.frame corresponding
to a locus is not 4 as expected.

\item Configuration files to be imported by the \texttt{readPanels()}
function may have more than one tabulation separator between
two data items in a way that could be different from one
line to another one. There is an example of such a case in the
test file \texttt{Prototype\_PowerPlex\_EP01\_Pa.txt} where
locus \texttt{D10S1248} and \texttt{D22S1045} are followed by a
single tabulation when all remaining loci are followed by two
tabulations.
This was a source of trouble during importation.
This is now fixed by preprocessing the input so that
all consecutive tabulations are replaced by a single one.
The above mentioned test file is now used as a quality control.

\end{itemize}

\subsection*{release 2.0-3}

\begin{itemize}

\item As pointed out on the seqinr diffusion list on 23-APR-2009
by Darren Obbard there was an obscure error message when function
\texttt{kaks()} was called with an alignment such that the number
of nucleotides was not a multiple of 3 after gap removal. This check
was partial as an alignment with out-of-frame gaps but with
a total number of gaps multiple of 3 was not detected.
The new behaviour is that if at least one non ACGT base is found in a
codon, then the whole codon is forced to a gap codon (\texttt{---}).
The documentation of the function has been clarified accordingly,
and a new alignment file \texttt{DarrenObbard.fasta} added in
the \texttt{sequences} folder to check this new behaviour.

\item Function \texttt{readBins()} is now more tolerant when there
is an extra column with possibly empty fields in data by forcing
the \texttt{fill} argument of \texttt{read.table()} function to
\texttt{TRUE}.

\item As pointed out by e-mail on 30-MAR-2009 by Yann Lesecque there 
was a bug in the \texttt{getTrans()} function: when applied to a 
list of sequences with all the same length the returned result 
was a matrix instead of a list. This is now fixed.

\item New utility functions \texttt{readPanels()} and \texttt{readBins()} 
to import data from GeneMapper configuration files. Four example files
are now in the \texttt{abif} folder.

\item Function \texttt{peakabif()} now returns the heights and
surfaces of peaks in addition to their location.

\item New utility function \texttt{al2bp()} to convert
a forensic microsatellite allele name into its length in base pairs.
Conventions used to name forensic microsatellite alleles (STR) are described
in Bar \emph{et al.} (1994) \cite{BarW1994}. 
The name \texttt{9.3} means for instance that there are
9 repetitions of the complete base oligomer and an incomplete 
repeat with 3 bp.

\end{itemize}

\subsection*{release 2.0-2}

\begin{itemize}

\item New ABIF format related functions: \texttt{plotabif()} to plot
electrophoregrams with optonial internal size standard and optional
allelic ladder, \texttt{peakabif()} to locate peaks in electrophoregrams,
\texttt{plotladder()} to display an observed allelic ladder.

\item New datasets \texttt{gs500liz} for size standards, \texttt{identifiler}
for allelic ladder names, \texttt{ECH} for allelic ladder raw data and
\texttt{JLO} for forensic genetic profile raw data. The last one is
now used as a quality check for the \texttt{read.abif()} function.

\item A new folder called \texttt{abif} has been created under the
\texttt{inst} folder. The purpose of this folder is to contain examples
of files in ABIF format so that the results of the \texttt{read.abif()} 
function can be checked against expected results for quality check.
It contains for now two duplicated genetic profiles and two allelic
ladders from the same batch experiment.

\end{itemize}

\subsection*{release 2.0-1}

\begin{itemize}

\item The useless \texttt{itemize} in the argument section of
  documentation file \texttt{stresc.Rd} is now deleted.

\item In function \texttt{words.pos()} the default value for parameter
  \texttt{extended} was changed from FALSE to TRUE to avoid warnings.

\item New experimental function \texttt{read.abif()} to import files
  in ABIF format (\texttt{*.fsa}, (\texttt{*.ab1}).

\end{itemize}

\subsection*{release 2.0-0}

\begin{itemize}

\item New draft chapter about making RISA \textit{in silico} added.

\item Objects from class \texttt{qaw} created after a call to the
  \texttt{query()} function have gained a new generic \texttt{print}
  method to focus on the most important information: number
  of sequences in the list, list type and the corresponding request.

\item Function \texttt{query()} now allows a missing \texttt{listname}
  argument. In this case, \texttt{list1} is used to store the
  result.

\item Function \texttt{autosocket()} has been changed to behave more
  friendly with outdated R versions. This is essentially a backward
  compatibility issue that will not be maintained in the future.
  The function \texttt{autosocket()} works hard to check that everything
  is OK with the last opened database, especially with the socket
  infos available in \texttt{banknameSocket\$socket} thru its 
  \texttt{summary()} generic. In old R versions (\textit{e.g.} 2.6.2) this
  was returning \texttt{socket} instead of \texttt{sockconn} for the
  class, yielding an error in seqinR 1.1-7. The old result is now
  allowed but a warning is issued.

\end{itemize}

The 2.0 series started in summer 2008 along with the moving
of the seqinr sources on R-forge.

\section*{1.1 series}

\subsection*{release 1.1-7}

\begin{itemize}

\item As suggested by Kurt Hornik two extra \texttt{cr} in the
  documentation file for \texttt{ec999} were deleted.

\item Function \texttt{read.fasta()} has gained four new arguments
  (\textit{viz.} \texttt{bfa}, \texttt{sizeof.longlong},
  \texttt{endian}, \texttt{apply.mask}) to read DNA binary fasta
  files in MAQ format. There is a new \texttt{ct.bfa} file
  in the \texttt{sequences} folder to check for the MAQ format
  reading.

\item New dataset \texttt{pK} for the values for the side chain of 
  charged amino acids from various sources compiled by Joanna Kiraga
  \cite{KiragaJ2008}.

\item Function \texttt{words.pos()} has gained new arguments that
  are passed to \texttt{regexpr()} including the dot-dot-dot
  argument in case of need in the future. The documentation has
  been modified to better explain the difference with the standard
  \texttt{gregexpr()} function.

\item As pointed by e-mail on 28 May 2008 by Kim Milferstedt a
  function to compute the consensus for a set of aligned sequences
  would be helpful. There is now a function \texttt{consensus()}
  aliased to \texttt{con()} for this. The input is either an object
  from class \texttt{alignment} or a matrix of characters. The
  output is either a consensus sequence (using the majority rule,
  the majority rule with a threshold, or IUPAC symbols for RNA
  and DNA sequences) or a profile, that is a matrix with the count 
  of each possible character at each position in the alignment.

\item In the documentation of the \texttt{read.alignment()} function
  a link was added to the \texttt{read.nexus()} function from the
  \texttt{ComPairWise} package \cite{ComPairWise}.

\item New function \texttt{bma()} to find the IUPAC symbol corresponding
  to a nucleic sequence.

\item New function \texttt{as.matrix.alignment()} to convert an
  alignment into a object of class \texttt{matrix}.

\item The encoding of line ends in the example file \texttt{test.mase}
  is now an unix-like one.

\item As pointed by e-mail on 31 May 2008 by Marie S{\'e}mon there was
  no convenient function to compute the Codon Adaptation Index \cite{CAI}.
  A new function \texttt{cai()} was introduced with the aim of reproducing
  exactly the results from the program \texttt{codonW} that was written
  by John Peden during his PhD thesis \cite{codonW} under the supervision 
  of P.M. Sharp (the most authorative source for CAI computation). A new 
  dataset \texttt{caitab} that was hard-encoded in \texttt{codonW} for
  the \texttt{w} values for some species (\textit{viz}
  \textit{Escherichia coli}, \textit{Bacillus subtilis},
  \textit{Saccharomyces cerevisiae}) was added. Care was taken to credit
  original sources. The \textit{E. coli} data that was uncredited is
  from \cite{CAI}. The \textit{B. subtilis} data that was uncredited
  is from \cite{ShieldsDC1987} (see the note of caution in \texttt{?caitab}
  before using this one directly to compute CAI in \textit{B. subtilis}). 
  The \textit{S. cerevisiae} data that was
  credited to \cite{SharpPM1991} dates back from \cite{CAI}. A new
  text file \texttt{scuco.txt} produced by \texttt{codonW} was added
  in the \texttt{sequences} folder to check that the CAI results from
  \texttt{cai()} are consistents with thoses from \texttt{codonW} version
  1.4.4 (03-MAR-2005). This legacy file is used in the example section of
  the \texttt{cai()} function.

\end{itemize}

\subsection*{release 1.1-6}

\begin{itemize}

\item The construct \texttt{get(getOption("device"))(width = 18, height = 11)}
  that was used in the example section for \texttt{data(prochlo)} is no
  more valid since \Rlogo{}~2.8.0 (fall 2008). The example has been
  restricted to work only with \texttt{X11}, \texttt{windows} and
  \texttt{quartz} devices.

\item As pointed by e-mail on 12 May 2008 by Indranuj Mukherjee there
  was a bug in the function \texttt{oriloc()}: when called with a
  \texttt{gbk = NULL} argument the function was trying to remove
  non-existent files, yielding an error. The bug has been fixed and
  the documentation of the function \texttt{oriloc()} has been
  extended to better explain how to use the arguments \texttt{seq.fasta}
  and \texttt{gbk}.

\item A reference to \cite{GautierC1985} was missing in the documentation
  of function \texttt{zscore()} for the codon model.

\item As suggested by e-mail on 11 Mar 2008 by Christian Gautier,
  the function \texttt{count()}
  has gained a new argument \texttt{by} to control the window step,
  allowing for instant to count dinucleotides in codon position III-I
  in a coding sequence. The example section of the function documentation
  has been extended to give an example of counting dinucleotides in
  position III-I.

\begin{Schunk}
\begin{Sinput}
 alldinuclIIIpI <- s2c("NNaaNatNttNtgNgtNtcNctNtaNagNggNgcNcgNgaNacNccNcaNN")
 (resIIIpI <- count(alldinuclIIIpI, word = 2, start = 2, by = 3))
\end{Sinput}
\begin{Soutput}
aa ac ag at ca cc cg ct ga gc gg gt ta tc tg tt 
 1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1 
\end{Soutput}
\begin{Sinput}
 stopifnot(all(resIIIpI == 1))
\end{Sinput}
\end{Schunk}

\item Function \texttt{reverse.align()} has gained two arguments
  \texttt{forceDNAtolower = TRUE} and \texttt{forceAAtolower = FALSE} 
  that are passed to the functions used to read the sequences.
  There is now a new dataset \texttt{revaligntest} used to check
  the result in the example section of \texttt{reverse.align()}.

\item As pointed by e-mail on 21 Feb 2008 by Oliver Keatinge Clay
  function \texttt{modifylist()} failed to scan in GenBank FEATURES
  annotation lines. There is now a new function called
  \texttt{prepgetannots()}, aliased to \texttt{pga()}, that allows
  to set up the annotation lines to be scanned. Called with default
  arguments, this function turns on all annotation lines for scan.
  This function can also be used to set up partly the annotation lines
  to be returned by \texttt{getAnnot()}.

\item Function \texttt{choosebank()} has gained four arguments
  (\texttt{server}, \texttt{blocking}, \texttt{open}, \texttt{encoding})
  that are passed to \texttt{socketConnection()}. The value of
  the argument \texttt{verbose} is now passed to \texttt{clientid()}
  which knows now how to handle it. The \texttt{encoding} argument
  was introduced to fix a localization bug on Mac OS X which symptom was a
  cryptic error message in \texttt{if (res[1] != "0") \{} after
  a call to \texttt{choosebank()}. The culprit was an
  \texttt{option(encoding = "latin1")} that was set up before
  the call to \texttt{choosebank()} who called \texttt{socketConnection()}
  with its default \texttt{encoding = getOption("encoding")},
  preventing \texttt{readLines()} to read from the socket.
  The bug was fixed by opening the socket with the native encoding,
  which is the current default.

\item As pointed by e-mail on 15 Jan 2008 by Stefanie Hartmann, the
  argument \texttt{frame} in function \texttt{count()} was misleading
  for someone with a molecular biology background. The argument has
  been replaced by \texttt{start}. The old argument name is maintained
  as an alias for backward compatibility. The example section has 
  been extended to
  give an example with the complete human mitochondrion sequence, the
  corresponding fasta file (\texttt{humanMito.fasta}) has been added in the 
  \texttt{sequences} directory.
\end{itemize}

\subsection*{release 1.1-5}

Minor release to fix mainly problems in the documentation.

\begin{itemize}

\item The argument section was empty in \texttt{autosocket.Rd}.

\item The details section was empty in \texttt{countfreelists.Rd} and
  \texttt{draw.oriloc.Rd}.

\item The value section was empty in \texttt{gbk2g2.Rd}. The corresponding
  function was changed to use a local file for the demo.

\item The description section was missing in \texttt{getFrag.Rd},
  \texttt{getLength.Rd}, \texttt{getName.Rd}, \texttt{getSequence.Rd}.

\item Documentation of the function \texttt{dia.bactgensize()} to plot 
  the distribution of bacterial genome size from GOLD data has been
  ammended to credit sources \cite{gold1, gold2, gold3, gold4}. It
  has gained a new argument \texttt{maxgensize} defaulting to 20000 
  to remove outliers. It has also gained a new argument \texttt{source}
  for the file to look for raw data, defaulting to an (outdated) local
  copy so that the function can be called even when there is no internet
  connection.

\end{itemize}

\subsection*{release 1.1-4 (10-Dec-2007)}

Minor release to fix problems found by Kurt Hornik.

\begin{itemize}

\item In the \texttt{DESCRIPTION} file \texttt{License: GPL (>= 2)} instead
  of \texttt{License: GPL version 2 or newer}.
 
\item The files \texttt{inst/doc/src/mainmatter/acnuc\_sockets.rnw .tex}
  with non-portable file names were changed to \texttt{acnucsocket.rnw}
  and \texttt{acnucsocket.tex}.

\end{itemize}

\subsection*{release 1.1-3}

\begin{itemize}

\item There is a new chapter to explain how to set up a
  local ACNUC server on Unix-like platforms.

\item New dataset \texttt{m16j} to make a GC skew plot as in
  \cite{LobryMBE96}.

\item New dataset \texttt{waterabs} giving the absorption of light 
  by water. This dataset was compiled by Palmeira \cite{PalmeiraL2007}
  from \cite{LitjensRA1999, QuickendenTI1980}.

\item Generic functions \texttt{getAnnot()}, \texttt{getFrag()},
  \texttt{getKeyword()}, \texttt{getLength()}, \texttt{getLocation()},
  \texttt{getName()}, \texttt{getSequence()} and \texttt{getTrans()}
  have gained methods to handle objects from class \texttt{list}
  and \texttt{qaw}.

\item Functions \texttt{getAttributsocket()} and \texttt{getNumber.socket()}
  are now deprecated, a warning is issued.

\item There is a new appendix in which all the examples protected
  by a \texttt{dontrun} statment are forced to be executed.

\item Function \texttt{read.fasta()} now supports comment lines
  starting by a semicolon character in FASTA files. An example
  of such a file is provided in \texttt{sequences/legacy.fasta}.
  The argument \texttt{File} is now deprecated. There is
  a new argument \texttt{seqonly} to import just the sequences
  without names, annotations and coercion attempts. There is
  a new argument \texttt{strip.desc} to remove the leading
  '>' character in annotations (as in function \texttt{readFASTA}
  from the Biostrings package \cite{Biostrings}). The FASTA file
  example \texttt{someORF.fsa} from Biostrings is also added
  for comparisons.

\item Function \texttt{GC()} has gained a new argument \texttt{NA.GC}
  defaulting to \texttt{NA} to say what should be returned when the
  GC content cannot be computed from data (for instance with a
  sequence like NNNNNNNNNNNN). The argument \texttt{oldGC} is now
  deprecated and a warning is issued. Functions \texttt{GC1()},
  \texttt{GC2()}, \texttt{GC3()} are now simple wrappers for the
  more general \texttt{GCpos()} function. The new argument \texttt{frame}
  allows to take the frame into account for CDS.

\item Function \texttt{read.alignment()} has gained a new argument
  \texttt{forceToLower} defaulting to TRUE to force lower case in
  the character of the sequence (this is for a smoother interaction
  with the package \texttt{ape}). The argument \texttt{File} is now
  deprecated and a warning is issued when used instead of \texttt{file}. 
  The example in the function \texttt{kaks()} has been corrected
  to avoid this warning when reading the example files.

\item New low level utility function \texttt{acnucclose()} and
  \texttt{quitacnuc()} to close an ACNUC server. These functions
  are called by \texttt{closebank()} so that a simple call to
  it should be enough.

\item New low level utility function \texttt{clientid()} to
  send the client ID to an ACNUC server.

\item New low level utility function \texttt{countfreelists()} to
  get the number of free lists available in an ACNUC server.

\item New low level utility function \texttt{knowndbs()} and its
  shortcut \texttt{kdb()} to get a description of databases 
  known by an ACNUC server.

\item New low level utility function \texttt{autosocket()} to get the
  socket connection to the last opened ACNUC database.

\item New function \texttt{countsubseqs()} to get the number of 
  subsequences in an ACNUC list.

\item New function \texttt{savelist()} to save sequence names or 
  accession numbers from an ACNUC list into a local file.

\item New function \texttt{ghelp()} to get help from an ACNUC server.

\item New function \texttt{modifylist()} to modify a previously existing 
  ACNUC list by selecting sequences either by length, either by date, 
  either for the presence of a given string in annotations.

\item New low level function \texttt{getlistate()} to ask for 
  information about an ACNUC list.

\item New low level function \texttt{setlistname()} to set the name of 
  a list from an ACNUC server.

\item New function \texttt{residuecount()} to count the total number 
  of residues (nucleotides or aminoacids) in all sequences of an ACNUC
  list of specified rank.

\item New function \texttt{isenum()} and its shortcut \texttt{isn()}
  to get the ACNUC number of a sequence from its name or accession
  number.

\item New function \texttt{prettyseq()} to get a text representation 
  of a sequence from an ACNUC server.

\item New function \texttt{gfrag()} to extract sequence identified by 
  name or by number from an ACNUC server.

\item The details of the socket connection are no more stored in
  the slot \texttt{socket} for objects of class \texttt{seqAcnucWeb}:
  this slot is now deleted. As a consequence, the argument
  \texttt{socket} in function \texttt{as.SeqAcnucWeb()} has been
  removed and there is now a new
  argument \texttt{socket = "auto"} in functions \texttt{getAnnot()},
  \texttt{getFrag()}, \texttt{geyKeyword()}, \texttt{getLocation()},
  and \texttt{getSequence()}. The default value \texttt{"auto"} means
  that the details of the socket connection are taken automatically
  when necessary from the last opened bank. The size of local lists
  of sequences is reduced by about a third now as compared to the
  previous version.

\item New function \texttt{print.seqAcnucWeb()} to print objects from
  class \texttt{seqAcnucWeb}.

\item Internal function \texttt{parser.socket()} has been optimized and
  is about four times faster now. This decreases the time needed by the
  \texttt{query()} function.

\end{itemize}

\subsection*{release 1.1-2}


\begin{itemize}

\item New function \texttt{trimSpace()} to remove leading and trailing
  spaces in string vectors.

\item Function \texttt{splitseq()} is no more based on \texttt{substring()},
  it is now more efficient for long sequences.

\item A sanity check test was added in the documentation file
  for the function \texttt{syncodons()}.

\item The way this manual is produced is now documented in the
  \texttt{doc/src/template/} folder.

\item A bug in function \texttt{oriloc()} was reported on 23 Jul 2007
  by Michael Kube: using directly genBank files was no more possible.
  The culprit was \texttt{gbk2g2()} that turns genBank files into
  glimmer files version 2 when \texttt{oriloc()} default is to use
  version 3 files. The \texttt{glimmer.version} argument is now forced
  to 2 when working with genBank files to fix this problem.
 
\item Function \texttt{zscore()} has now a new argument
  \texttt{exact} (which is only effective for the option
  \texttt{model = base}). This argument, when set to 
  \texttt{TRUE} allows for the exact analytical computation 
  of the zscore under this model, instead of the approximation for
  large sequences. It is set to \texttt{FALSE} by
  default for backward compatibility.
\end{itemize}

\subsection*{release 1.1-1}


\begin{itemize}
\item A bug was reported by Sylvain Mousset on 14 Jul 2007 in function
\texttt{dist.alignment()}: when called with sequences in lower case
letters, some sequences were modified. This should no more be the case:
\begin{Schunk}
\begin{Sinput}
 ali <- list(nb = 4, nam = c("speciesA", "speciesB", "speciesC", 
     "speciesD"), seq = c("ACGT", "acgt", "ACGT", "ACGT"))
 class(ali) <- "alignment"
 print(ali$seq)
\end{Sinput}
\begin{Soutput}
[1] "ACGT" "acgt" "ACGT" "ACGT"
\end{Soutput}
\begin{Sinput}
 print(dist.alignment(ali))
\end{Sinput}
\begin{Soutput}
         speciesA speciesB speciesC
speciesB        0                  
speciesC        0        0         
speciesD        0        0        0
\end{Soutput}
\begin{Sinput}
 print(ali$seq)
\end{Sinput}
\begin{Soutput}
[1] "ACGT" "acgt" "ACGT" "ACGT"
\end{Soutput}
\end{Schunk}

\item The CITATION file has been updated so that now \texttt{citation("seqinr")} returns the full complete reference for the package seqinR.

\item Non ASCII characters in documentation
(*.Rd) files have been removed. Declaration of the encoding as 
latin1 when necessary is now present. 
The updated documentation files are: \texttt{
dinucl.Rd,
gb2fasta.Rd,
get.ncbi.Rd,
lseqinr.Rd,
n2s.Rd,
prochlo.Rd,
s2c.Rd,
SeqAcnucWeb.Rd,
SeqFrag.Rd,
toyaa.Rd,
words.pos.Rd,
words.Rd,
zscore.Rd}.

\item Function \texttt{GC()} and by propagation functions \texttt{GC1()},
\texttt{GC2()} and \texttt{GC3()} have gained a new argument \texttt{oldGC}
allowing to compute the G+C content as in releases up to 1.0-6 included.
The code has been also modified to avoid divisions by zero with very
small sequences.

\item New function \texttt{rot13()} that returns the ROT-13 encoding of
a string of characters.

\end{itemize}

\section*{1.0 series}

\subsection*{release 1.0-7}

\begin{itemize}

\item A new \emph{experimental} function \texttt{extractseqs()} to download
sequences thru zlib compressed sockets from an ACNUC server is released. 
Preliminary tests suggest that working with about 100,000 CDS is possible with 
a home ADSL connection. See the manual for some \texttt{system.time()}
examples.

\item As pointed by e-mail on 16 Nov 2006 by Emmanuel Prestat the URL
used in \texttt{dia.bactgensize()} was no more available, this has been fixed
in the current version.

\item As pointed by e-mail on 16 Nov 2006 by Guy Perri{\`e}re, the
function \texttt{oriloc()} was no more compatible with glimmer\footnote{
Glimmer is a program to predict coding sequences in microbial genomes \cite{SalzbergSL1998, DelcherAL1999}.
} 3.0
outputs. The function has gained a new argument \texttt{glimmer.version}
defaulting to 3, but the value 2 is still functional for backward compatibility
with old glimmer outputs.

\item As pointed by e-mail on 24 Oct 2006 by Lionel Guy 
(\url{http://pbil.univ-lyon1.fr/seqinr/seqinrhtmlannuel/03/0089.html})
there was no default value for the \texttt{as.string} argument in
the \texttt{getSequence.SeqFastadna()}. A default \texttt{FALSE}
value is now present for backward compatibility with older code.

\item New utility vectorized function \texttt{stresc()}  to escape \LaTeX~special 
characters present in a string.

\item New low level function \texttt{readsmj()} available.

\item A new function \texttt{readfirstrec()} to get 
the record count of the specified ACNUC index file is now available.

\item Function \texttt{getType()} called without arguments will now use
the default ACNUC database to return available subsequence types.

\item Function \texttt{read.alignment()} now also accepts \texttt{file} in
addition to \texttt{File} as argument.

\item A new function \texttt{rearranged.oriloc()} is available. This
method, based on \texttt{oriloc()}, can be used to detect the effect of
the replication mechanism on DNA base composition asymmetry, in
prokaryotic chromosomes.

\item New function \texttt{extract.breakpoints()}, used to extract
breakpoints in rearranged nucleotide skews. This function uses the
\texttt{segmented} package to define the position of the breakpoints.

\item New function \texttt{draw.rearranged.oriloc()} available, to plot
nucleotide skews on artificially rearranged prokaryotic chromosomes.

\item New function \texttt{gbk2g2.euk()} available. Similarly to 
\texttt{gbk2g2()}, this function extracts the coding sequence annotations
from a GenBank format file. This function is specifically designed for
eukaryotic sequences, \textit{i.e.} with introns. The output file will contain
the coordinates of the exons, along with the name of the CDS to which
they belong.

\item After an e-mail by Marcelo Bertalan on 26 Mar 2007, a bug in
\texttt{oriloc()} when the \texttt{gbk} argument was \texttt{NULL}
was found and fixed by Anamaria Nec\c{s}ulea.

\item Functions \texttt{translate()} and \texttt{getTrans()} have gained
a new argument \texttt{NAstring} to represent untranslatable amino-
acids, defaulting to character "X".

\item There was a typo for the total number of printed bases in
the ACNUC books \cite{GautierC1982a, GautierC1982b} : 474,439 should be 526,506.

\item Function \texttt{invers()} has been deleted.

\item Functions \texttt{translate()}, \texttt{getTrans()} and \texttt{comp()} have gained a 
new argument \texttt{ambiguous} defaulting to FALSE allowing to handle 
ambiguous bases. If TRUE, ambiguous bases are taken into account so that 
for instance GGN is translated to Gly in the standard genetic code.

\item New function \texttt{amb()} to return the list of nucleotide matching 
a given IUPAC nucleotide symbol.

\item Function \texttt{count()} has gained a new argument \texttt{alphabet}
so that oligopeptides counts are now possible. Thanks to Gabriel Valiente
for this suggestion. The functions \texttt{zscore()}, \texttt{rho()} and
\texttt{summary.SeqFastadna()} have also an argument \texttt{alphabet} which
is forwarded to \texttt{count()}.

\end{itemize}

\subsection*{release 1.0-6}

Release 1.0-6 is a minor release to fix a problem found and solved by Kurt Hornik
(namely a change from \texttt{SET\_ELEMENT} to \texttt{SET\_STRING\_ELT}
in C code for \texttt{s2c()} in file \texttt{util.c}). The few changes are
as follows.

\begin{itemize}

\item More typographical option for the output \LaTeX~table of \texttt{tablecode()}
are now available to outline deviations from the standard genetic code (see example in the
appendix "genetic codes" of the manual).

\item A new dataset \texttt{aaindex} extracted from the aaindex database
\cite{aaindex1, aaindex2, aaindex3} is now available. It contains a list
of 544 physicochemical and biological properties for the 20 amino-acids

\item The default value for argument \texttt{dia} is now \texttt{FALSE}
in function \texttt{tablecode()}.

\item The example code for \texttt{data(chargaff)} has been changed.

\end{itemize}

\subsection*{release 1.0-5}

\begin{itemize}

\item A  new function \texttt{dotPlot()} is now available.

\item A new function \texttt{crelistfromclientdata()} is now available to
create a list on the server from a local file of sequence names, sequence
accession numbers, species names, or keywords names.

\item A new function \texttt{pmw()} to compute the molecular weight of
a protein is now available.

\item A new function \texttt{reverse.align()} contributed by Anamaria Nec\c{s}ulea
is now available to align CDS at the protein level and then reverse translate this at
the nucleic acid level from a \texttt{clustalw} output. This can be done on the fly
if \texttt{clustalw} is available on your platform.

\item An undocumented behavior was reported by Guy Perri{\`e}re for \texttt{uco()}
when computing RSCU on sequences where an amino-acid is missing. There is
now a new argument \texttt{NA.rscu} that allows the user to force the
missing values to his favorite magic value.

\item There was a bug in \texttt{read.fasta()}: some sequence names were
truncated, this is now fixed (thanks to Marcus G. Daniels for pointing this).
In order to be more consistent with standard functions such as \texttt{read.table()}
or \texttt{scan()}, the file argument starts now with a lower case letter (\texttt{file})
in function \texttt{read.fasta()}, but the old-style \texttt{File} is still
functional for forward-compatibility. There is a new logical argument in \texttt{read.fasta()}
named \texttt{as.string} to allow sequences to be returned as strings instead of
vector of single characters. The automatic conversion of DNA sequences into
lower case letters can now be disabled with the new logical argument
\texttt{forceDNAtolower}. It is also possible to disable the automatic attributes
settings with the new logical argument \texttt{set.attributes}.

\item A new function \texttt{write.fasta()} is now available.

\item The function \texttt{kaks()} now forces character in sequences to upper case.
This default behavior can be neutralized in order to save time by setting the 
argument \texttt{forceUpperCase} to \texttt{FALSE}.

\end{itemize}

\subsection*{release 1.0-4}

\begin{itemize}
\item The scaling factor $n_{\bullet\bullet}$ was missing in equation \ref{chi}.
\item The files \texttt{louse.fasta}, \texttt{louse.names}, \texttt{gopher.fasta}, \texttt{gopher.names}
and \texttt{ortho.fasta} that were used for examples in the previous version of this document are
no more downloaded from the internet since they are now distributed in the \texttt{sequences/} folder
of the package.
\item An example of synonymous and non synonymous codon usage analysis was added to the
vignette along with two toy data sets (\texttt{toyaa} and \texttt{toycodon}).
\item A FAQ section was added to the vignette.
\item A bug in \texttt{getAnnot()} when the number of lines was zero is now fixed.
\item There is now a new argument, \texttt{latexfile}, in \texttt{tablecode()} to export genetic codes
tables in a \LaTeX~document, for instance table \ref{../tables/code3.tex} and table \ref{../tables/code4.tex} here.
\item There is now a new argument, \texttt{freq}, in \texttt{count()}
  to compute word frequencies instead of counts.
\item Function \texttt{splitseq()} has been entirely rewritten to improve speed.
\item Functions computing the G+C content: \texttt{GC(), GC1(), GC2(),
  GC3()} were rewritten to improve speed, and their document files
  were merged to facilitate usage.
\item The following new functions have been added:
\begin{itemize}
\item \texttt{syncodons()} returns all synonymous codons for a given
  codon. Argument \texttt{numcode} specifies the desired genetic code.
\item \texttt{ucoweight()} returns codon usage bias on a sequence as
  the number of synonymous codons present in the sequence for each
  amino acid.
\item \texttt{synsequence()} generates a random coding sequence which
      is synonymous to a given sequence and has a chosen codon usage
      bias.
\item \texttt{permutation()} generates a new sequence from a given
  sequence, while maintaining some constraints from the given sequence
  such as nucleotide frequency, codon usage bias, ...
\item \texttt{rho()} computes the rho statistic on dinucleotides as
  defined in \cite{Karlin}.
\item \texttt{zscore()} computes the zscore statistic on dinucleotides
  as defined in \cite{UV}.
\end{itemize}
\item Two datasets (\texttt{dinucl} and \texttt{prochlo}) were added
  to illustrate these new functions.

\end{itemize}

\subsection*{release 1.0-3}

\begin{itemize}
\item The new package maintainer is Dr. Simon Penel, PhD, who has now a fixed position in the
laboratory that issued \seqinr~(\texttt{penel@biomserv.univ-lyon1.fr}). Delphine Charif was
successful too to get a fixed position in the same lab, with now a different research task (but who knows?).
Thanks to the close vicinity of our pioneering maintainers the transition was sweet. The DESCRIPTION
file of the \seqinr{} package has been updated to take this into account.

\item The reference paper for the package is now \textit{in press}. We do not have the full
reference for now, you may use \texttt{citation("seqinr")} to check if it is complete now:
\begin{Schunk}
\begin{Sinput}
 citation("seqinr")
\end{Sinput}
\begin{Soutput}
To cite seqinR in publications use:

  Charif, D. and Lobry, J.R. (2007)

A BibTeX entry for LaTeX users is

  @incollection{,
    author = {D. Charif and J.R. Lobry},
    title = {Seqin{R} 1.0-2: a contributed package to the {R} project for statistical computing devoted to biological sequences retrieval and analysis.},
    booktitle = {Structural approaches to sequence evolution: Molecules, networks, populations},
    year = {2007},
    editor = {U. Bastolla, M. Porto, H.E. Roman and M. Vendruscolo},
    series = {Biological and Medical Physics, Biomedical Engineering},
    pages = {207-232},
    address = {New York},
    publisher = {Springer Verlag},
    note = {{ISBN :} 978-3-540-35305-8},
  }

Note that the orginal article updated is available in the
/Users/lobry/seqinr/pkg.Rcheck/seqinr/doc/ folder in PDF format
\end{Soutput}
\end{Schunk}

\item There was a bug when sending a \texttt{gfrag} request to the server for long (Mb range) 
sequences. The length argument was converted to scientific notations that are not understand by the
server. This is now corrected and should work up the the Gb scale.

\item The \texttt{query()} function has been improved by de-looping list element info request,
there are now download at once which is much more efficient. For example, a query from a
researcher-home ADSL connection with a list with about 1000 elements was 60 seconds and
is now only 4 seconds (\textit{i.e.} 15 times faster now).

\item A new parameter \texttt{virtual} has been added to \texttt{query()} 
so that long lists can stay on the server without trying to download
them automatically. A query like \texttt{query(s\$socket,"allcds","t=cds", virtual = TRUE)} is 
now possible.

\item Relevant genetic codes and frames are now automatically propagated.

\item \Seqinr{}~sends now its name and version number to the server.

\item Strict control on ambiguous DNA base alphabet has been relaxed.

\item Default value for parameter \texttt{invisible} of function \texttt{query()} is now \texttt{TRUE}.

\end{itemize}



\section*{Session Informations}

\begin{scriptsize}

This part was compiled under the following \Rlogo{}~environment:

\begin{itemize}\raggedright
  \item R version 2.10.0 (2009-10-26), \verb|i386-apple-darwin8.11.1|
  \item Locale: \verb|fr_FR.UTF-8/fr_FR.UTF-8/fr_FR.UTF-8/C/C/C|
  \item Base packages: base, datasets, graphics, grDevices, grid,
    methods, stats, utils
  \item Other packages: ade4~1.4-13, ape~2.4, grImport~0.4-4,
    MASS~7.3-3, quadprog~1.4-11, seqinr~2.0-7, tseries~0.10-21,
    XML~2.6-0, xtable~1.5-5, zoo~1.5-8
  \item Loaded via a namespace (and not attached): gee~4.13-14,
    lattice~0.17-26, nlme~3.1-96
\end{itemize}
There were two compilation steps:

\begin{itemize}
  \item \Rlogo{} compilation time was: Sun Nov 15 11:55:18 2009
  \item \LaTeX{} compilation time was: \today
\end{itemize}

\end{scriptsize}

% END - DO NOT REMOVE THIS LINE


%%%%%%%%%%%%  BIBLIOGRAPHY %%%%%%%%%%%%%%%%%
\clearpage
\addcontentsline{toc}{section}{References}
\bibliographystyle{plain}
\bibliography{../config/book}
\end{document}
